Add docs for array/object types and refine section strucutre (#10347)
mlflow/mlflow: docs/source/models.rst

include the following additional metadata about model inputs, outputs and params that can be used by
downstream tooling:

* :ref:`Model Inference Params <inference-params>` - description of params used for model inference.
* :ref:`Model Signature <model-signature>` - description of a model's inputs, outputs and parameters.
* :ref:`Model Signature <model-signature>` - description of a model's inputs, outputs and additional parameters for inference.
* :ref:`Model Input Example <input-example>` - example of a valid model input.

.. _inference-params:

> Model Inference Params

Author: Note from offline converstaion:

`Inference parameter` might not be a concept limited to signature, as we can directly pass it to `predict()` function. Instead of locating it under signature section as a sub section, we can have a top-level section for it, so that the signature section can just describe basic usage and linke to it for details.

Author: @serena-ruan  Probably not gonna address this in this PR, as it's hard to find alternative place to put inference params section in, with current doc strucutre. Ideally, we should have some introductory section that shows how to run inference with MLflow model, and describe about inference params concept there. 

Re-structuring Models page is must-do anyway, it's by-far too long. For now I prioritize Tracking page as I already started on similar restructuring for it. Then I can come back once we find the best shape there, and find better place for inference params.

Reviewer: @BenWilson2 What is our plan with spliting models.rst file? And new folder structures settled down?

Author: Note from stand-up: We are trying to port contents out from `models.rst` into children pages and make `models.rst` just a hub and overview page. I'll raise another PR to split model signature (+probably a few more sections like schema enforcement) and put inference param section in appropriate location on the new pages.

Column-based Signature
~~~~~~~~~~~~~~~~~~~~~~
This signature is typically used for traditional machine learning models that take tabular data as input such as Pandas 
DataFrame. Column-based schemas are a sequence of (optionally) named columns with data type. Each column-based input and 
output is represented by a type corresponding to one of :ref:`supported data types <supported-data-types-column>` and an 
optional name. Input columns can also be marked as ``optional``, indicating whether they are required as input to the 
model or can be omitted. 

The following example displays the model signature for a classification model trained on the 
`Iris dataset <https://archive.ics.uci.edu/ml/datasets/iris>`_. The input has 4 named, numeric columns and 
1 named, optional string column. The output is an unnamed integer specifying the predicted class.

.. code-block:: yaml

  signature:
      inputs: '[
        {"name": "sepal length (cm)", "type": "double"}, 
        {"name": "sepal width(cm)",   "type": "double"}, 
        {"name": "petal length (cm)", "type": "double"}, 
        {"name": "petal width (cm)",  "type": "double"}, 

>         {"name": "class",             "type": "string",  "optional": "true"}

Reviewer 2: Can we remove `class`?

Reviewer 2: ```suggestion
```

>         {"name": "class",             "type": "string",  "optional": "true"}

Reviewer 2: ```suggestion
        {"name": "class",             "type": "string",  "required": "true"}
```

do we need to change optional to required?

Reviewer: Oh we might need to add 'required': true to every line

Author: > do we need to change optional to required?

Ah yes, will do.

Btw I also think we should remove that field. It's the target class we want to predict:p https://archive.ics.uci.edu/dataset/53/iris

      ]'
      outputs: '[{"type": "integer"}]'
      params: null

.. _supported-data-types-column:

Supported Data Types
""""""""""""""""""""
Column-based signature supports :py:class:`MLflow DataType <mlflow.types.DataType>` that defines following data primitives,

> * string
> * integer
> * boolean
> * double
> * float
> * long
> * datetime

Reviewer 2: ```suggestion
* string
* integer
* long
* float
* double
* boolean
* datetime
```

.. code-block:: python

    from mlflow.models import infer_signature

    sign = infer_signature(model_input={"long_col": 1, "str_col": "a", "bool_col": True})

>     print(sign.input)

Reviewer 2: ```suggestion
    print(sign.inputs)
```

Reviewer 2: can we also fix other `sign.input`?

    # [
    #   'long_col': long (required),
    #   'str_col': string (required),
    #   'bool_col': boolean (required)
    # ]

>     ## You can also infer signature from Pandas DataFrame. Column with None value will be `optional`

Reviewer 2: ```suggestion
    ## You can also infer signature from Pandas DataFrame. Column with None will be inferred as `optional`
```

    sign = infer_signature(model_input=pd.DataFrame({"col": [1.0, 2.0, None]}))
    print(sign.input)
    # [ 'col': long (optional) ]

> Column-based signature also support composite data types of these primitives,

Reviewer 2: ```suggestion
Column-based signature also support composite data types of these primitives:
```

* Array (list, numpy arrays)
* Object (dictionary)

.. code-block:: python

    # Array (from list)
    sign = infer_signature(model_input={"list_col": ["a", "b", "c"]})
    print(sign.input)
    # [ 'list_col': Array(string) (required) ]

> 

Reviewer: I'm thinking that what we want to highlight here might be the result `[ 'list_col': Array(string) (required) ]` instead of the code 🤔 I feel like above yaml part of showing signature example is much clearer than this. WDYT? Though this is not a blocker, It explains what we want to show, just wondering what people thinks about it :D

Author: Yeah I agree that yaml is the best way to print schema and comment is a bit ugly...
One challenge is that we still want to show python code for input sample. Ideally we can show python/yaml side-by-side with nice style, but currently it doesn't work nicely.

Reviewer: What about a table showing 'inputs' and 'inferred schema' ?

Author: Table works nicely indeed.

![Screenshot 2023-11-15 at 16 57 44](https://github.com/mlflow/mlflow/assets/31463517/25fdbc4e-cd05-4fbb-adb7-0ce3420a22ec)

A few pain points exist, but probably the best we can do for now (hopefully it becomes easier after md migration lol)
1. Only for single row apprently, if there is more than one, it gives gray background that doesn't look great. 
2. Super annoying to write .rst table with code block. It's fine for a few samples, but definitely not > 10.

<img width="1127" alt="Screenshot 2023-11-15 at 17 17 02" src="https://github.com/mlflow/mlflow/assets/31463517/897134e7-89fb-4cd2-bc4c-75709f22a3f3">

Reviewer: Thanks for exploring this!! Though I was simply referring to text in the table lol, but this definitely looks better. We could also keep the current one migrate to tables afterwards :)

    # Array (from numpy array)
    sign = infer_signature(model_input={"np_col": np.array([1, 2, 3])})
    print(sign.input)
    # [ 'np_col': Array(long) (required) ]

    # Object
    sign = infer_signature(
        model_input={"object_col": {"long_prop": 1, "array_prop": ["a", "b", "c"]}}
    )
    print(sign.input)
    # ['object_col': {
    #     long_prop: long (required),
    #     str_prop: Array(string) (required)
    #  } (required)]

    # Combination of array / object is also supported
    sign = infer_signature(
        model_input={
            "complex_col": [
                {
                    "prop_a": [["a", "b", "c"], ["d", "e", "f"]],
                    "prop_b": 0.1,
                },
                {
                    "prop_a": [["g"]],
                    "prop_b": 0.5,
                },
            ]
        }

>     )

Reviewer 2: Is `sign.inputs` missing here?

    # [
    #    'complex_col': Array({
    #      prop_a: Array(Array(string)) (required),
    #      prop_b: double (required)
    #     }) (required)
    #  ]

.. note::
    Nested arrays can contain an empty list. In such case, the schema will be inferred

>     from the other elements of the list, assuming they have homogeneous types.

Reviewer 2: Can we add a small example for this?

Author: Sure, will add!

Model Signature Types
~~~~~~~~~~~~~~~~~~~~~

A model signature consists of three schemas **(1) inputs**, **(2)outputs**, and **(3) inference params**. The
former two simply refer to the input and output of the model, while the inference params is a set of additional params 
that are passed to the model at inference time. 

Input and output schemas can be either column-based or tensor-based. 

Column-based Signature
""""""""""""""""""""""""""""""

> 

Reviewer: Remove this blank line?

This signature is typically used for traditional machine learning models that take tabular data as input such as Pandas 
DataFrame. Column-based schemas are a sequence of (optionally) named columns with data type. Each column-based input and 

> output is represented by a type corresponding to one of :ref:`supported data types <supported-data-types>` and an 

Author: Yup, I added those types in `supported data types` section

optional name. Input columns can also be marked as ``optional``, indicating whether they are required as input to the 
model or can be omitted. 

The following example displays the model signature for a classification model trained on the 
`Iris dataset <https://archive.ics.uci.edu/ml/datasets/iris>`_. The input has 4 named, numeric columns and 
1 named, optional string column. The output is an unnamed integer specifying the predicted class.

.. code-block:: yaml

  signature:
      inputs: '[
        {"name": "sepal length (cm)", "type": "double"}, 
        {"name": "sepal width(cm)",   "type": "double"}, 
        {"name": "petal length (cm)", "type": "double"}, 
        {"name": "petal width (cm)",  "type": "double"}, 
        {"name": "class",             "type": "string",  "optional": "true"}
      ]'
      outputs: '[{"type": "integer"}]'
      params: null

Tensor-based Signature
""""""""""""""""""""""""""""""

> This signature is typically used for deep learning models that take tensors as input such as images, speech, etc. 

Reviewer: It's not just for deep learning models, could we just use `models`?

Author: That's right, probably sth like `models that take tensors such as imagesa and speech as input, as represented by deep learning`.

            return list(params.values())

    params = {"temperature": 0.5, "suppress_tokens": [101, 102]}
    # params' default values are saved with ModelSignature
    signature = infer_signature(["input"], params=params)

    with mlflow.start_run():
        model_info = mlflow.pyfunc.log_model(
            python_model=MyModel(), artifact_path="my_model", signature=signature
        )

    loaded_model = mlflow.pyfunc.load_model(model_info.model_uri)

    # Not passing params -- predict with default values
    loaded_predict = loaded_model.predict(["input"])
    assert loaded_predict == [0.5, [101, 102]]

    # Passing some params -- add default values
    loaded_predict = loaded_model.predict(["input"], params={"temperature": 0.1})
    assert loaded_predict == [0.1, [101, 102]]

    # Passing all params -- override
    loaded_predict = loaded_model.predict(
        ["input"], params={"temperature": 0.5, "suppress_tokens": [103]}
    )
    assert loaded_predict == [0.5, [103]]

.. note::
    When validating param values, the values will be converted to python native types.
    For example, ``np.float32(0.1)`` will be converted to ``float(0.1)``.

.. _supported-data-types:

> Supported Data Types in Model Signature

Reviewer: ```suggestion
Supported Data Types in ColSpec
```

Reviewer: Maybe we could move this section inside `Column-based signature` and explain this is the data types supported inside such schema.

Author: One trickly thing here is it has overlap with inference params.

Author: Actually found this is great idea while writing, thanks for the suggestion!

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Currently, :py:class:`MLflow DataType <mlflow.types.DataType>` defines following data primitives for model signature.

Reviewer: We specifically just added it inside ColSpec, so might want to update this sentence.

* string
* integer
* boolean
* double
* float
* long
* datetime

> Model signature also supports composite data types as follows,

Reviewer: ```suggestion
ColSpec also supports composite data types as follows,
```

Author: Like above comment, I'm a bit afraid of using internal term like `ColSpec` is confusing. Probably we can have this kind of subsections to clarify the difference in supported types?

```
Columns
---------

Columns in column-based model signature supports following types 

* string
* integer
* ...
 
also supports composite ....

Tensors
--------

`numpy` data types blah blah

Params
--------

Primitives plus 1-D array
```

> * Array

Reviewer: ```suggestion
* Array (list, numpy arrays)
```

* Object (a.k.a. dictionary in python)

> * Nested Arrays and Objects e.g. ``Array(Object(Propertiy("a", Array(DataType.integer))))``

Reviewer: We might not need to explicitly list this here as it's also Array. We can explain in Array section about what internal types we support.

> Here are some examples of model inputs and inferred signature

Reviewer: Can we add different sections for `Array` and `Object`, and write example for them separately? Also maybe we don't need `model_output` as we just need to show example about inferring signature from some data. We could construct each example as 1. infer_signature 2. directly constructing the signature.

Author: > Can we add different sections for Array and Object, and write example for them separately? 

Yeah I think it's doable, to have `Array` `Object` subsections.

> Also maybe we don't need model_output

Got it. The reason I added model_output is that I felt doing `signature.input` everytime is a bit ugly, no strong preference tho.

> We could construct each example as 1. infer_signature 2. directly constructing the signature.

Hmm with same reason above, I don't think we should surface internal class like `ColSpec`, `Schema` too much for users. We have another section below for constructing schema manually (explicitly).

      outputs: '[{"shape": [-1, 10], "dtype": "float32"}]'
      params: null

.. _supported-data-types-tensor:

Supported Data Types
""""""""""""""""""""

> Tensor-based schemas supports `numpy data types <https://numpy.org/devdocs/user/basics.types.html>`_.

Reviewer 2: ```suggestion
Tensor-based schemas support `numpy data types <https://numpy.org/devdocs/user/basics.types.html>`_.
```

                 "shape": [-1]
            }
        ]'

The values for inference params are passed to the model at inference time in the form of a dictionary 
and each param value will be validated against the corresponding param type defined in the model signature. 
The following example demonstrates how to define params in a model signature and how to use them for model inference.

.. code-block:: python

    import mlflow
    from mlflow.models import infer_signature

    class MyModel(mlflow.pyfunc.PythonModel):
        def predict(self, ctx, model_input, params):
            return list(params.values())

    params = {"temperature": 0.5, "suppress_tokens": [101, 102]}
    # params' default values are saved with ModelSignature
    signature = infer_signature(["input"], params=params)

    with mlflow.start_run():
        model_info = mlflow.pyfunc.log_model(
            python_model=MyModel(), artifact_path="my_model", signature=signature
        )

    loaded_model = mlflow.pyfunc.load_model(model_info.model_uri)

    # Not passing params -- predict with default values
    loaded_predict = loaded_model.predict(["input"])
    assert loaded_predict == [0.5, [101, 102]]

>     # Passing some params -- add default values

Reviewer 2: ```suggestion
    # Passing some params -- override passed-in params
```

    loaded_predict = loaded_model.predict(["input"], params={"temperature": 0.1})
    assert loaded_predict == [0.1, [101, 102]]

>     # Passing all params -- override

Reviewer 2: ```suggestion
    # Passing all params -- override all params
```
